Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp

```diff
@@ -0,0 +1,529 @@
+{
+    WTF::Persistence::Encoder encoder;
+    encoder << currentVersion;
+    encoder << key;
+    encoder << WallTime { };
+    encoder << computeSHA1(headerData.span(), salt);
+    encoder << (uint64_t)headerData.size();
+    encoder << bodyHash;
+    encoder << (uint64_t)bodyData.size();
+    encoder << isBodyInline;
+    encoder.encodeChecksum();
+
+    auto metaData = Vector<uint8_t> { encoder.buffer(), encoder.bufferSize() };
+    Vector<uint8_t> result;
+    result.appendVector(metaData);
+    result.appendVector(headerData);
+
+    StringBuilder sb;
+    for (auto& data : bodyData) {
+        sb.append(data);
+        sb.append(",");
+    }
+    
+    if (isBodyInline)
+        result.appendVector(bodyData);
+
+    return result;
+}
+
+void CacheStorageDiskStore::writeRecords(Vector<CacheStorageRecord>&& records, WriteRecordsCallback&& callback)
+{
+    Vector<String> recordFiles;
+    Vector<Vector<uint8_t>> recordDatas;
+    Vector<Vector<uint8_t>> recordBlobDatas;
+    for (auto&& record : records) {
+        recordFiles.append(recordFilePath(record.info.key));
+        auto bodyData = encodeRecordBody(record);
+        auto bodyHash = computeSHA1(bodyData.span(), m_salt);
+        bool shouldCreateBlob = shouldStoreBodyAsBlob(bodyData);
+        auto headerData = encodeRecordHeader(WTFMove(record));
+        auto recordData = encodeRecord(record.info.key, headerData, !shouldCreateBlob, bodyData, bodyHash, m_salt);
+        recordDatas.append(WTFMove(recordData));
+        if (!shouldCreateBlob)
+            bodyData = { };
+        recordBlobDatas.append(WTFMove(bodyData));
+    }
+
+    m_ioQueue->dispatch([this, protectedThis = Ref { *this }, recordFiles = WTFMove(recordFiles), recordDatas = WTFMove(recordDatas), recordBlobDatas = WTFMove(recordBlobDatas), callback = WTFMove(callback)]() mutable {
+        bool result = true;
+        // FIXME: we should probably stop writing and revert changes when result becomes false.
+        for (size_t index = 0; index < recordFiles.size(); ++index) {
+            auto recordFile = recordFiles[index];
+            auto recordData = recordDatas[index];
+            auto recordBlobData = recordBlobDatas[index];
+            FileSystem::makeAllDirectories(FileSystem::parentPath(recordFile));
+            if (!recordBlobData.isEmpty())  {
+                if (FileSystem::overwriteEntireFile(recordBlobFilePath(recordFile), Span { recordBlobData.data(), recordBlobData.size() }) == -1) {
+                    result = false;
+                    continue;
+                }
+            }
+            if (FileSystem::overwriteEntireFile(recordFile, Span { recordData.data(), recordData.size() }) == -1)
+                result = false;
+        }
+
+        m_callbackQueue->dispatch([protectedThis = Ref { *this }, result, callback = WTFMove(callback)]() mutable {
```

Comment:
Why protectedThis? If this is to make sure we do not delete this in the io queue, should we instead do protectedThis = WTFMove(protectedThis)?

Reply:
Sure.